File: webflux-dialogue/src/main/java/com/study/webflux/rag/infrastructure/credit/document/UserCreditDocument.java
package com.study.webflux.rag.infrastructure.credit.document;

import java.time.Instant;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Version;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

@Document(collection = "user_credits")
public record UserCreditDocument(
	@Id String id,
	@Indexed(unique = true) String userId,
	long balance,
	@Version long version,
	Instant updatedAt
) {
	public static UserCreditDocument fromDomain(
		com.study.webflux.rag.domain.credit.model.UserCredit credit) {
		return new UserCreditDocument(
			credit.userId().value(),
			credit.userId().value(),
			credit.balance(),
			credit.version(),
			Instant.now());
	}

	public com.study.webflux.rag.domain.credit.model.UserCredit toDomain() {
		return new com.study.webflux.rag.domain.credit.model.UserCredit(
			com.study.webflux.rag.domain.dialogue.model.UserId.of(userId),
			balance,
			version);
	}
}

File: infrastructure/src/main/kotlin/com/miyou/app/infrastructure/credit/document/UserCreditDocument.kt
package com.miyou.app.infrastructure.credit.document

import com.miyou.app.domain.credit.model.UserCredit
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Version
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document
import java.time.Instant

@Document(collection = "user_credits")
data class UserCreditDocument(
    @Id val id: String,
    @Indexed(unique = true) val userId: String,
    val balance: Long,
    @Version val version: Long,
    val updatedAt: Instant,
) {
    companion object {
        fun fromDomain(credit: UserCredit): UserCreditDocument =
            UserCreditDocument(
                credit.userId(),
                credit.userId(),
                credit.balance(),
                credit.version(),
                Instant.now(),
            )
    }

    fun toDomain(): UserCredit =
        UserCredit(
            userId,
            balance,
            version,
        )
}
